src/gluonts/trainer/model_averaging.py

```diff
@@ -0,0 +1,250 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+# Standard library imports
+from typing import Any, Dict, List, Tuple
+import glob
+import json
+
+# Third-party imports
+import numpy as np
+import mxnet as mx
+
+# First-party imports
+from gluonts.core.component import validated
+
+EPOCH_INFO_STRING = "epoch-info"
+
+
+def save_epoch_info(tmp_path: str, epoch_info: dict) -> None:
+    r"""
+    Writes the current epoch information into a json file in the model path.
+
+    Parameters
+    ----------
+    tmp_path
+        Temporary base path to save the epoch info.
+    epoch_info
+        Epoch information dictionary containing the parameters path, the epoch number and the tracking metric value.
+
+    Returns
+    -------
+    None
+    """
+
+    with open("{}-{}.json".format(tmp_path, EPOCH_INFO_STRING), "w") as f:
+        json.dump(epoch_info, f)
+
+
+class AveragingStrategy:
+    @validated()
+    def __init__(
+        self,
+        num_models: int = 5,
+        metric: str = "score",
+        maximize: bool = False,
+    ):
+        r"""
+        Parameters
+        ----------
+        num_models
+            Number of model checkpoints to average.
+        metric
+            Metric which is used to average models.
+        maximize
+            Boolean flag to indicate whether the metric should be maximized or minimized.
+        """
+        self.num_models = num_models
+        self.metric = metric
+        self.maximize = maximize
+
+    def apply(self, model_path: str) -> str:
+        r"""
+        Averages model parameters of serialized models based on the selected model strategy and metric.
+        IMPORTANT: Depending on the metric the user might want to minimize or maximize. The maximize flag has to be
+        chosen appropriately to reflect this.
+
+        Parameters
+        ----------
+        model_path
+            Path to the models directory.
+
+        Returns
+        -------
+        Path to file with the averaged model.
+        """
+        checkpoints = self.load_top_checkpoints(model_path)
```

Comment:
I think this can be more flexible here, by not pre-selecting the top n. For example how would we pick the last n epochs even if they didn't have the best scores?

At this point we should get a list with all available checkpoints and their metdata (score, epoch etc) and then call a method `select_checkpoints` with this data. The method can then return the weights and checkpoint paths that should be averaged.

The `select_checkpoints` method could then differ for different strategies. So we could then have `SelectNBest`, `AverageLastN` etc.

Does that make sense?

Reply:
The assumption here was that you always want to select the best and not some random epochs. I kept this part (selecting top checkpoints) common is all classes since it was a common assumption, regardless of how each class does the averaging later. I am not sure if it is useful to average the last N and not the best N, but still we can include this option. 

This increases a bit the complexity of defining things. I see two options here:

1) The class should take both some info on which epochs to select and how to average them, e.g. 

`MyStrategy(num_models=N, select='last', average='mean')`

which basically removes the need for subclasses but creates an `if ... elif` block for checking all the options of `select` and `average` (unless we actually create classes for selection and averaging and pass instances which makes things more complex in defining things e.g. `MyStrategy(num_models=N, select=SelectLast() average=MeanAvg())`). 

2) Create specific classes for each option but then the number of classes grows too much, e.g.

`MeanAvgLast(num_models=N)  -> implicit select='last', average='mean'`
`MeanAvgBest(num_models=N )-> implicit  select='best', average='mean'`
`MetricValueAvgBest(num_models=N)-> implicit  select='best', average='metric'`

Any suggestions on what is better?

Reply:
Yeah, I think keeping it flexible is best. For instance it may be good to pick the best epoch and average the n surrounding epochs. Specific classes and the `select_checkpoints` method is the most flexible approach I think.